#include "qtbattleview.h"
#include "ui_qtbattleview.h"
#include "../model/Pet.hpp"
#include <QMessageBox>
#include <QScrollBar>
#include <QTextCursor>
#include <random>

//æ„é€ å‡½æ•°å®ç°
QtBattleView::QtBattleView(Player* player, QWidget* parent) :
    QWidget(parent)
    , ui(new Ui::QtBattleView)
    , _player(player)
    , _battleStarted(false)
    , _autoBattle(false)
    , _pendingDisplayUpdate(false)
{
    ui->setupUi(this);

    _autoTimer = new QTimer(this);

    // ä¿®å¤é‡å¤è¿æ¥é—®é¢˜
    setupConnections();

    ui->start_font->setAttribute(Qt::WA_TransparentForMouseEvents);
    ui->auto_play_font->setAttribute(Qt::WA_TransparentForMouseEvents);
    ui->forward_font->setAttribute(Qt::WA_TransparentForMouseEvents);
    ui->start->setAttribute(Qt::WA_TransparentForMouseEvents);
    ui->auto_play->setAttribute(Qt::WA_TransparentForMouseEvents);
    ui->forward->setAttribute(Qt::WA_TransparentForMouseEvents);

    setupPetDisplays();
}

void QtBattleView::setupConnections()
{
    // å…ˆæ–­å¼€æ‰€æœ‰å¯èƒ½çš„è¿æ¥
    disconnect(ui->start_button, nullptr, this, nullptr);
    disconnect(ui->auto_play_button, nullptr, this, nullptr);
    disconnect(ui->forward_button, nullptr, this, nullptr);
    disconnect(_autoTimer, nullptr, this, nullptr);

    // ä½¿ç”¨UniqueConnectionç¡®ä¿åªè¿æ¥ä¸€æ¬¡
    connect(_autoTimer, &QTimer::timeout, this, &QtBattleView::onAutoStep, Qt::UniqueConnection);
    connect(ui->start_button, &QPushButton::clicked, this, &QtBattleView::on_start_button_clicked, Qt::UniqueConnection);
    connect(ui->auto_play_button, &QPushButton::clicked, this, &QtBattleView::on_auto_play_button_clicked, Qt::UniqueConnection);
    connect(ui->forward_button, &QPushButton::clicked, this, &QtBattleView::on_forward_button_clicked, Qt::UniqueConnection);
}

//å¯¹æˆ˜å¼€å§‹å®ç°
void QtBattleView::startNewBattle()
{
    _battleStarted = false;
    _autoBattle = false;
    _pendingDisplayUpdate = false;
    _autoTimer->stop();

    // ç”ŸæˆAIå¯¹æ‰‹
    generateAITeam(_player->getRound());

    // å»¶è¿Ÿæ›´æ–°æ˜¾ç¤º
    QTimer::singleShot(0, this, [this]() {
        updateBattleDisplay();
    });

    // æ¢å¤è‡ªåŠ¨æ’­æ”¾å›¾æ ‡å’Œæ–‡å­—çš„åŸå§‹çŠ¶æ€
    ui->auto_play->setPixmap(QPixmap(":/else/photo/Refresh.png"));
    ui->auto_play_font->setText("<html><head/><body><p><span style=\" font-size:11pt; font-weight:700;\">è‡ªåŠ¨æ’­æ”¾</span></p></body></html>");

    // ç§»é™¤é€æ˜åº¦æ•ˆæœ
    ui->start->setGraphicsEffect(nullptr);

    // è®¾ç½®é€æ˜åº¦æ•ˆæœ
    QGraphicsOpacityEffect *iconEffect_1 = new QGraphicsOpacityEffect(this);
    QGraphicsOpacityEffect *iconEffect_2 = new QGraphicsOpacityEffect(this);
    iconEffect_1->setOpacity(0.5);
    iconEffect_2->setOpacity(0.5);
    ui->auto_play->setGraphicsEffect(iconEffect_1);
    ui->forward->setGraphicsEffect(iconEffect_2);
    ui->auto_play_font->setStyleSheet("color: rgba(255, 255, 255, 0.5);");
    ui->forward_font->setStyleSheet("color: rgba(255, 255, 255, 0.5);");

    // è´´ä¸Šç©å®¶æ•°æ®
    ui->heart_left_num->setText(QString::number(_player->getLives()));
    ui->heart_right_num->setText(QString::number(_player->getLives()));
    ui->hourglass_left_num->setText(QString::number(_player->getRound()));
    ui->hour_glass_right_num->setText(QString::number(_player->getRound()));
    ui->trophy_left_num->setText(QString::number(_player->getPrize()));
    ui->trophy_right_num->setText(QString::number(_player->getPrize()));

    // æ¢å¤æ–‡å­—åŸå§‹æ ·å¼
    ui->start_font->setStyleSheet("color: white;");

    ui->start_button->setEnabled(true);
    ui->start_button->setCursor(Qt::PointingHandCursor);
    ui->auto_play_button->setEnabled(false);
    ui->auto_play_button->setCursor(Qt::ArrowCursor);
    ui->forward_button->setEnabled(false);
    ui->forward_button->setCursor(Qt::ArrowCursor);

}

//ç”ŸæˆAIé˜Ÿä¼å®ç°
void QtBattleView::generateAITeam(int difficulty)
{
    _aiTeam.clear();

    std::random_device rd;
    std::mt19937 gen(rd());

    // æ ¹æ®éš¾åº¦ç”Ÿæˆå¯¹æ‰‹ï¼ˆ2-5åªå® ç‰©ï¼‰
    int petCount = std::min(5, 2 + difficulty / 2);

    for (int i = 0; i < petCount; ++i)
    {
        // éšæœºé€‰æ‹©å® ç‰©ç±»å‹
        int petType = std::uniform_int_distribution<>(0, 7)(gen);

        std::unique_ptr<Pet> pet;
        switch (petType)
        {
        case 0: pet = std::make_unique<Cat>(4, 3, 1, 1); break;
        case 1: pet = std::make_unique<Ant>(2, 1, 1, 1); break;
        case 2: pet = std::make_unique<Fish>(3, 2, 1, 1); break;
        case 3: pet = std::make_unique<Cricket>(1, 2, 1, 1); break;
        case 4: pet = std::make_unique<Swan>(3, 1, 1, 2); break;
        case 5: pet = std::make_unique<Flamingo>(3, 2, 1, 2); break;
        case 6: pet = std::make_unique<Hedgehog>(3, 1, 1, 2); break;
        case 7: pet = std::make_unique<Peacock>(2, 5, 1, 3); break;
        }

                // æ ¹æ®éš¾åº¦æå‡å±æ€§
        int levelBonus = difficulty / 3;
        for (int j = 0; j < levelBonus; ++j)
        {
            pet->gainExperience(2);
        }

        _aiTeam.push_back(std::move(pet));
    }

}

void QtBattleView::setupPetDisplays()
{
    // æ¸…ç©ºå®¹å™¨
    _playerPetLabels.clear();
    _aiPetLabels.clear();
    _playerPetStatusLabels.clear(); // æ¸…ç©ºçŠ¶æ€å›¾æ ‡ç­¾
    _aiPetsStatusLabels.clear();    // æ¸…ç©ºçŠ¶æ€å›¾æ ‡ç­¾
    _playerAttackLabels.clear();    // æ–°å¢: æ¸…ç©ºå®¹å™¨
    _playerHPLabels.clear();        // æ–°å¢: æ¸…ç©ºå®¹å™¨
    _aiAttackLabels.clear();        // æ–°å¢: æ¸…ç©ºå®¹å™¨
    _aiHPLabels.clear();            // æ–°å¢: æ¸…ç©ºå®¹å™¨

            // ä½¿ç”¨UIä¸­é¢„ç•™çš„å® ç‰©ä½ç½® (é¡ºåºä¸ºä»å³åˆ°å·¦: 5, 4, 3, 2, 1)

    // ç©å®¶å® ç‰©ï¼ˆä»å³åˆ°å·¦ï¼‰
    _playerPetLabels.append(ui->your_pet_5);
    _playerPetLabels.append(ui->your_pet_4);
    _playerPetLabels.append(ui->your_pet_3);
    _playerPetLabels.append(ui->your_pet_2);
    _playerPetLabels.append(ui->your_pet_1);

    // ç©å®¶çŠ¶æ€æ ‡ç­¾ (Status Image)
    _playerPetStatusLabels.append(ui->your_pet_5_status); // å‡è®¾æ‚¨çš„çŠ¶æ€å›¾ç‰‡æ ‡ç­¾å‘½åä¸º your_status_X
    _playerPetStatusLabels.append(ui->your_pet_4_status);
    _playerPetStatusLabels.append(ui->your_pet_3_status);
    _playerPetStatusLabels.append(ui->your_pet_2_status);
    _playerPetStatusLabels.append(ui->your_pet_1_status);

    // ç©å®¶å±æ€§æ ‡ç­¾ (Attack)
    _playerAttackLabels.append(ui->your_Attack_5); // å‡è®¾æ‚¨çš„æ”»å‡»åŠ›æ•°å€¼æ ‡ç­¾å‘½åä¸º your_attack_X
    _playerAttackLabels.append(ui->your_Attack_4);
    _playerAttackLabels.append(ui->your_Attack_3);
    _playerAttackLabels.append(ui->your_Attack_2);
    _playerAttackLabels.append(ui->your_Attack_1);

    // ç©å®¶å±æ€§æ ‡ç­¾ (HP)
    _playerHPLabels.append(ui->your_HP_5); // å‡è®¾æ‚¨çš„ç”Ÿå‘½å€¼æ•°å€¼æ ‡ç­¾å‘½åä¸º your_hp_X
    _playerHPLabels.append(ui->your_HP_4);
    _playerHPLabels.append(ui->your_HP_3);
    _playerHPLabels.append(ui->your_HP_2);
    _playerHPLabels.append(ui->your_HP_1);


    // AI å® ç‰©ï¼ˆä»å·¦åˆ°å³: 1, 2, 3, 4, 5)
    _aiPetLabels.append(ui->opponents_pet_1);
    _aiPetLabels.append(ui->opponents_pet_2);
    _aiPetLabels.append(ui->opponents_pet_3);
    _aiPetLabels.append(ui->opponents_pet_4);
    _aiPetLabels.append(ui->opponents_pet_5);

    // AI çŠ¶æ€æ ‡ç­¾ (Status Image)
    _aiPetsStatusLabels.append(ui->opponents_pet_1_status); // å‡è®¾æ‚¨çš„çŠ¶æ€å›¾ç‰‡æ ‡ç­¾å‘½åä¸º opponents_status_X
    _aiPetsStatusLabels.append(ui->opponents_pet_2_status);
    _aiPetsStatusLabels.append(ui->opponents_pet_3_status);
    _aiPetsStatusLabels.append(ui->opponents_pet_4_status);
    _aiPetsStatusLabels.append(ui->opponents_pet_5_status);

    // AI å±æ€§æ ‡ç­¾ (Attack)
    _aiAttackLabels.append(ui->opponents_Attack_1); // å‡è®¾æ‚¨çš„æ”»å‡»åŠ›æ•°å€¼æ ‡ç­¾å‘½åä¸º opponents_attack_X
    _aiAttackLabels.append(ui->opponents_Attack_2);
    _aiAttackLabels.append(ui->opponents_Attack_3);
    _aiAttackLabels.append(ui->opponents_Attack_4);
    _aiAttackLabels.append(ui->opponents_Attack_5);

    // AI å±æ€§æ ‡ç­¾ (HP)
    _aiHPLabels.append(ui->opponents_HP_1); // å‡è®¾æ‚¨çš„ç”Ÿå‘½å€¼æ•°å€¼æ ‡ç­¾å‘½åä¸º opponents_hp_X
    _aiHPLabels.append(ui->opponents_HP_2);
    _aiHPLabels.append(ui->opponents_HP_3);
    _aiHPLabels.append(ui->opponents_HP_4);
    _aiHPLabels.append(ui->opponents_HP_5);
}

//æ›´æ–°å•ä¸ªå® ç‰©å›¾ç‰‡åŠå±æ€§å®ç°
void QtBattleView::updatePetDisplay(int index, bool isPlayer, const Pet* pet)
{

    QLabel* petLabel = isPlayer ? _playerPetLabels.value(index) : _aiPetLabels.value(index);
    QLabel* statusLabel = isPlayer ? _playerPetStatusLabels.value(index) : _aiPetsStatusLabels.value(index);
    QLabel* attackLabel = isPlayer ? _playerAttackLabels.value(index) : _aiAttackLabels.value(index);
    QLabel* hpLabel = isPlayer ? _playerHPLabels.value(index) : _aiHPLabels.value(index);

    if (!petLabel || !statusLabel || !attackLabel || !hpLabel) return;

    if (!pet) {
        petLabel->hide();    // å® ç‰©æ­»äº¡æˆ–ä½ç½®ä¸ºç©ºæ—¶ï¼Œæ¸…é™¤å¹¶éšè—æ‰€æœ‰ç›¸å…³æ ‡ç­¾
        statusLabel->hide(); // éšè— status å›¾ç‰‡
        attackLabel->hide(); // éšè—æ”»å‡»åŠ›æ•°å€¼
        hpLabel->hide();     // éšè—ç”Ÿå‘½å€¼æ•°å€¼
        return;
    }

    // å® ç‰©å­˜åœ¨æ—¶ï¼š
    petLabel->show();
    statusLabel->show(); // æ˜¾ç¤º status å›¾ç‰‡
    attackLabel->show(); // æ˜¾ç¤ºæ”»å‡»åŠ›æ•°å€¼
    hpLabel->show();     // æ˜¾ç¤ºç”Ÿå‘½å€¼æ•°å€¼

    // æ›´æ–°å® ç‰©å›¾ç‰‡ (ä¿æŒä¸å˜)
    QPixmap originalPixmap(QString(":/Pet/photo/Pet/%1.png").arg(pet->getName()));

    // å¦‚æœæ˜¯AIçš„å® ç‰©ï¼Œè¿›è¡Œæ°´å¹³ç¿»è½¬
    if (isPlayer) {
        QTransform transform;
        transform.scale(-1, 1);  // æ°´å¹³ç¿»è½¬
        originalPixmap = originalPixmap.transformed(transform);
    }

    petLabel->setPixmap(originalPixmap);
    petLabel->setScaledContents(true); // å¯ç”¨å†…å®¹è‡ªé€‚åº”

    // æ›´æ–° Attack å’Œ HP æ•°å€¼
    attackLabel->setText(QString::number(pet->getAttack()));
    hpLabel->setText(QString::number(pet->getHP()));

    // æ›´æ–° Tooltip (ä¿æŒä¸å˜)
    QString tooltip = QString("%1\nHP: %2\nATK: %3")
                          .arg(QString::fromStdString(pet->getName()))
                          .arg(pet->getHP())
                          .arg(pet->getAttack());

    if (pet->hasArmor()) tooltip += "\nğŸ›¡ï¸æŠ¤ç”²";
    if (pet->hasMelonShield()) tooltip += "\nğŸ‰è¥¿ç“œæŠ¤ç›¾";
    if (pet->canRevive()) tooltip += "\nğŸ„å¤æ´»";

    petLabel->setToolTip(tooltip);
}

//æ›´æ–°æˆ˜æ–—åœºæ™¯å®ç°
//æ›´æ–°æˆ˜æ–—åœºæ™¯å®ç°
void QtBattleView::updateBattleDisplay()
{
    if (!_player) return;

    clearHighlights();

    // ç¡®å®šé˜Ÿä¼æ•°æ®æº
    const std::vector<std::unique_ptr<Pet>>* aiTeamPtr;
    int playerTeamSize;

    if (_battleStarted) {
        // ========== çŠ¶æ€ 1: æˆ˜æ–—å·²å¼€å§‹ (ä½¿ç”¨ BattleEngine å®æ—¶æ•°æ®) ==========
        aiTeamPtr = &_battleEngine.getPlayer2Team();
        const auto& playerTeam = _battleEngine.getPlayer1Team();
        playerTeamSize = playerTeam.size();

        // 1. æ›´æ–°ç©å®¶å® ç‰©æ˜¾ç¤º
        for (int i = 0; i < _playerPetLabels.size(); ++i) {
            if (i < static_cast<int>(playerTeamSize)) {
                const Pet* pet = playerTeam[i].get();
                updatePetDisplay(i, true, pet); // updatePetDisplay ç°åœ¨å¤„ç†æ‰€æœ‰æ˜¾ç¤ºé€»è¾‘
            }
            else {
                updatePetDisplay(i, true, nullptr); // éšè—æ‰€æœ‰ç›¸å…³æ ‡ç­¾
            }
        }

    } else {
        // ========== çŠ¶æ€ 2: æˆ˜æ–—å°šæœªå¼€å§‹ (ä½¿ç”¨ Player å•†åº—é˜Ÿä¼å’Œ _aiTeam æˆå‘˜) ==========
        aiTeamPtr = &_aiTeam;
        playerTeamSize = _player->getPetCount();

        // 1. æ›´æ–°ç©å®¶å® ç‰©æ˜¾ç¤º (ä½¿ç”¨ _player å¯¹è±¡çš„é˜Ÿä¼)
        for (int i = 0; i < _playerPetLabels.size(); ++i) {
            if (i < playerTeamSize) {
                Pet* pet = _player->getPetAt(i);
                updatePetDisplay(i, true, pet); // updatePetDisplay ç°åœ¨å¤„ç†æ‰€æœ‰æ˜¾ç¤ºé€»è¾‘
            }
            else {
                updatePetDisplay(i, true, nullptr); // éšè—æ‰€æœ‰ç›¸å…³æ ‡ç­¾
            }
        }
    }

            // 2. æ›´æ–° AI å® ç‰©æ˜¾ç¤º (ç»Ÿä¸€ä½¿ç”¨ aiTeamPtr)
    for (int i = 0; i < _aiPetLabels.size(); ++i) {
        if (i < static_cast<int>(aiTeamPtr->size())) {
            const Pet* pet = (*aiTeamPtr)[i].get();
            updatePetDisplay(i, false, pet); // updatePetDisplay ç°åœ¨å¤„ç†æ‰€æœ‰æ˜¾ç¤ºé€»è¾‘
        }
        else {
            updatePetDisplay(i, false, nullptr); // éšè—æ‰€æœ‰ç›¸å…³æ ‡ç­¾
        }
    }

    setUpdatesEnabled(true);
    update();
}


//é«˜äº®ä¸‰æ­¥èµ°
void QtBattleView::highlightAttacker(int index, bool isPlayer)
{
    clearHighlights();

    auto& labels = isPlayer ? _playerPetLabels : _aiPetLabels;
    //å¾…å®ç°
    //if (index >= 0 && index < labels.size()) {
    //    labels[index]->setStyleSheet("border: 4px solid gold;");
    //}
}

void QtBattleView::highlightDefender(int index, bool isPlayer)
{
    auto& labels = isPlayer ? _playerPetLabels : _aiPetLabels;
    //å¾…å®ç°
    // if (index >= 0 && index < labels.size()) {
    //     labels[index]->setStyleSheet("border: 4px solid orange;");
    // }
}

void QtBattleView::clearHighlights()
{
   //å¾…å®ç°
   // æ¢å¤é»˜è®¤æ ·å¼
   // for (auto* label : _playerPetLabels) {
   //     label->setStyleSheet("border: 2px solid green;");
   // }
   // for (auto* label : _aiPetLabels) {
   //     label->setStyleSheet("border: 2px solid red;");
   // }
}

QtBattleView::~QtBattleView()
{
    delete ui;
}

//æŒ‰å¼€å§‹æŒ‰é’®
void QtBattleView::on_start_button_clicked()
{
    _battleStarted = true;

    // åˆ›å»ºç©å®¶å® ç‰©çš„å‰¯æœ¬ç”¨äºæˆ˜æ–—
    std::vector<std::unique_ptr<Pet>> playerPetsCopy;
    for (int i = 0; i < _player->getPetCount(); ++i) {
        Pet* originalPet = _player->getPetAt(i);
        if (originalPet) {
            // åˆ›å»ºå® ç‰©å‰¯æœ¬ï¼ˆéœ€è¦Petç±»æœ‰æ‹·è´æ„é€ å‡½æ•°ï¼‰
            auto petCopy = std::make_unique<Pet>(*originalPet);
            playerPetsCopy.push_back(std::move(petCopy));
        }
    }

    // åˆ›å»ºAIå® ç‰©çš„å‰¯æœ¬ç”¨äºæˆ˜æ–—
    std::vector<std::unique_ptr<Pet>> aiPetsCopy;
    for (size_t i = 0; i < _aiTeam.size(); ++i) {
        if (_aiTeam[i]) {
            qDebug() << "å¤åˆ¶AIå® ç‰©" << i << ":" << _aiTeam[i]->getName().c_str();

            // ä½¿ç”¨æ‹·è´æ„é€ å‡½æ•°åˆ›å»ºå‰¯æœ¬
            auto petCopy = std::make_unique<Pet>(*_aiTeam[i]);
            aiPetsCopy.push_back(std::move(petCopy));
        }
    }

    qDebug() << "ç©å®¶å® ç‰©å¤åˆ¶å®Œæˆï¼Œæ•°é‡:" << playerPetsCopy.size();
    qDebug() << "AIå® ç‰©å¤åˆ¶å®Œæˆï¼Œæ•°é‡:" << aiPetsCopy.size();

    // åˆå§‹åŒ–æˆ˜æ–—å¼•æ“
    _battleEngine.initialize(playerPetsCopy, aiPetsCopy, _player);

    // è®¾ç½®äº‹ä»¶å›è°ƒ
    _battleEngine.setEventCallback([this](const BattleEvent& event) {
        onBattleEvent(event);
    });

    // è§¦å‘æˆ˜æ–—å¼€å§‹äº‹ä»¶
    _battleEngine.startBattleManual();

    ui->start_button->setEnabled(false);
    ui->start_button->setCursor(Qt::ArrowCursor);
    QGraphicsOpacityEffect *iconEffect = new QGraphicsOpacityEffect(this);
    iconEffect->setOpacity(0.5); // 50% é€æ˜åº¦
    ui->start->setGraphicsEffect(iconEffect);
    ui->start_font->setStyleSheet("color: rgba(255, 255, 255, 0.5);");

    ui->auto_play->setGraphicsEffect(nullptr);
    ui->forward->setGraphicsEffect(nullptr);
    ui->auto_play_font->setStyleSheet("color: white");
    ui->forward_font->setStyleSheet("color: white");
    ui->auto_play_button->setEnabled(true);
    ui->auto_play_button->setCursor(Qt::PointingHandCursor);
    ui->forward_button->setEnabled(true);
    ui->forward_button->setCursor(Qt::PointingHandCursor);
}

//æŒ‰è‡ªåŠ¨æ’­æ”¾æŒ‰é’®
void QtBattleView::on_auto_play_button_clicked()
{
    if (_autoBattle)
    {
        // åœæ­¢è‡ªåŠ¨æˆ˜æ–—
        _autoBattle = false;
        _autoTimer->stop();
        qDebug() << "åœæ­¢è‡ªåŠ¨æ’­æ”¾";

        // æ”¹å˜è‡ªåŠ¨æ’­æ”¾å›¾æ ‡åŠæ–‡æœ¬
        ui->auto_play->setPixmap(QPixmap(":/else/photo/Refresh.png"));
        ui->auto_play_font->setText("<html><head/><body><p><span style=\" font-size:11pt; font-weight:700;\">è‡ªåŠ¨æ’­æ”¾</span></p></body></html>");
        // æ¢å¤æŒ‰é’®é€æ˜åº¦
        ui->forward->setGraphicsEffect(nullptr);
        ui->forward_font->setStyleSheet("color: white;");
        ui->forward_button->setEnabled(true);
        ui->forward_button->setCursor(Qt::PointingHandCursor);
    }
    else
    {
        // å¼€å§‹è‡ªåŠ¨æˆ˜æ–—
        _autoBattle = true;

        // è®¾ç½®æŒ‰é’®åŠé€æ˜æ•ˆæœ
        QGraphicsOpacityEffect *iconEffect_2 = new QGraphicsOpacityEffect(this);
        iconEffect_2->setOpacity(0.5);
        ui->forward->setGraphicsEffect(iconEffect_2);
        ui->forward_font->setStyleSheet("color: rgba(255, 255, 255, 0.5);");
        ui->forward_button->setEnabled(false);
        ui->forward_button->setCursor(Qt::ArrowCursor);

        // è‡ªåŠ¨æ’­æ”¾å›¾æ ‡ä¸æ–‡å­—æ”¹å˜
        ui->auto_play->setPixmap(QPixmap(":/else/photo/Pause.png"));
        ui->auto_play_font->setText("<html><head/><body><p><span style=\" font-size:11pt; font-weight:700;\">æš‚åœ</span></p></body></html>");

        _autoTimer->start(1500);  // æ¯1.5ç§’æ‰§è¡Œä¸€æ­¥
        // ç«‹å³æ‰§è¡Œç¬¬ä¸€æ­¥
        qDebug() << "ç«‹å³æ‰§è¡Œç¬¬ä¸€æ­¥";
        onAutoStep();
    }
}

//æŒ‰å•æ­¥æ‰§è¡ŒæŒ‰é’®
void QtBattleView::on_forward_button_clicked()
{
    // ç›´æ¥æ‰§è¡Œæˆ˜æ–—æ­¥éª¤
    bool hasMore = _battleEngine.executeSingleStep();

    if (!hasMore)
    {
        // æˆ˜æ–—ç»“æŸ
        BattleResult result = _battleEngine.getResult();
        emit battleFinished(result);

        ui->auto_play_button->setEnabled(false);
        ui->auto_play_button->setCursor(Qt::ArrowCursor);
        ui->forward_button->setEnabled(false);
        ui->forward_button->setCursor(Qt::ArrowCursor);
    }

    // å»¶è¿Ÿæ›´æ–°æ˜¾ç¤ºï¼Œé¿å…é˜»å¡
    QTimer::singleShot(10, this, [this]() {
        updateBattleDisplay();
    });
}

//å®ç°è‡ªåŠ¨æ‰§è¡Œ
void QtBattleView::onAutoStep()
{
    on_forward_button_clicked();

    if (!_battleEngine.isInBattle())
    {
        _autoTimer->stop();
        _autoBattle = false;
    }
}

//å®ç°å¯¹æˆ˜
void QtBattleView::onBattleEvent(const BattleEvent& event)
{
    // ç›´æ¥å¤„ç†äº‹ä»¶ï¼Œä½†æ ‡è®°éœ€è¦æ›´æ–°æ˜¾ç¤º
    switch (event.type)
    {
    case BattleEventType::BattleStart:
        break;

    case BattleEventType::TurnStart:
        highlightAttacker(event.attackerIndex, event.isPlayer1);
        highlightDefender(event.defenderIndex, !event.isPlayer1);
        _pendingDisplayUpdate = true;
        break;

    case BattleEventType::Attack:
        break;

    case BattleEventType::TakeDamage:
        _pendingDisplayUpdate = true;
        break;

    case BattleEventType::PetDeath:
        _pendingDisplayUpdate = true;
        break;

    case BattleEventType::SkillTrigger:
        _pendingDisplayUpdate = true;
        break;

    case BattleEventType::BattleEnd:
        clearHighlights();
        _pendingDisplayUpdate = true;
        break;
    }

    // å»¶è¿Ÿæ›´æ–°æ˜¾ç¤ºï¼Œé¿å…é¢‘ç¹é‡ç»˜å¯¼è‡´å¡é¡¿
    if (_pendingDisplayUpdate)
    {
        _pendingDisplayUpdate = false;
        QTimer::singleShot(10, this, [this]() {
            updateBattleDisplay();
        });
    }
}
